src/Features/Core/Portable/MetadataAsSource/MetadataAsSourceHelpers.cs

```diff
@@ -96,28 +103,244 @@ public static async Task<Location> GetLocationInGeneratedSourceAsync(SymbolKey s
             return null;
         }
 
-        public static bool TryGetImplementationAssemblyPath(string referenceDllPath, [NotNullWhen(true)] out string? implementationDllPath)
+        /// <summary>
+        /// Uses various heuristics to try to find the implementation assembly for a reference assembly without
+        /// loading 
+        /// </summary>
+        public static bool TryFindImplementationAssemblyPath(string referencedDllPath, [NotNullWhen(true)] out string? implementationDllPath)
         {
+            if (TryNugetLibToRef(referencedDllPath, out implementationDllPath))
+                return true;
+
+            if (TryTargetingPackToSharedSdk(referencedDllPath, out implementationDllPath))
+                return true;
+
             implementationDllPath = null;
+            return false;
+
+            static bool TryNugetLibToRef(string referenceDllPath, [NotNullWhen(true)] out string? implementationDllPath)
+            {
+                implementationDllPath = null;
+
+                // For some nuget packages if the reference path has a "ref" folder in it, then the implementation assembly
+                // will be in the corresponding "lib" folder.
+                var start = referenceDllPath.IndexOf(@"\ref\");
+                if (start == -1)
+                    return false;
+
+                var pathToTry = referenceDllPath.Substring(0, start) +
+                                @"\lib\" +
+                                referenceDllPath.Substring(start + 5);
+
+                if (IOUtilities.PerformIO(() => File.Exists(pathToTry)))
+                {
+                    implementationDllPath = pathToTry;
+                    return true;
+                }
 
-            // For some nuget packages if the reference path has a "ref" folder in it, then the implementation assembly
-            // will be in the corresponding "lib" folder.
-            // TODO: Support more cases, like SDK references: https://github.com/dotnet/sdk/issues/12360
-            var start = referenceDllPath.IndexOf(@"\ref\");
-            if (start == -1)
                 return false;
+            }
+
+            static bool TryTargetingPackToSharedSdk(string referencedDllPath, [NotNullWhen(true)] out string? implementationDllPath)
+            {
+                implementationDllPath = null;
+
+                // eg. C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.5\ref\net6.0\Foo.dll
+                var parts = referencedDllPath.Split(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
+                if (parts is not [.., "packs", var packName, var packVersion, "ref", _, var dllFileName])
```

Comment:
>parts [](http://example.com/codeflow?start=20&length=5)

nice pattern usage

Reply:
If we were on .NET 6 it would be `[.. var basePath, ` and be even better